### ISSUE TITLE
ncpa plugin is blocked if any programs are started in background

### ISSUE THREAD

**Issue by corkyman1** [NagiosEnterprises/ncpa#714]
There was a Nagios case 'ncpa plugin hangs  [#845059]' that describes the conditions and the circumvention.

When a plugin starts a program in the background and then ends the ncpa agent does not return, e.g. curl command executing the   plugin does not end. It ends only when the background process ends.  There is a circumvention for this problem:
instead of starting the background process like this:
 _program_ &
it should be called like this:
 _program_  > /dev/null 2>&1 & disown

It feels like the ncpa agent does not give control back unless all the child processes end, even if some are background processes.
At the very least it should be documented as a limitation as it requires a special coding of the plugin script.

**Comment by ccztux** [comment]
I have tested this with a python script emulating ncpa calling plugins and i can not reproduce that behaviour. Can you please provide an example?
During testing i found another issue with subprocesses opened via ncpa. If a plugin starts some background jobs, these background jobs can not be killed via ```Popen.kill()``` method. I will provide a fix for this via a pull request.

**Comment by corkyman1** [comment]
Here is my post to Sean who was working my case #845059. It has the code to reproduce the problem (I only removed commented out lines).

corkyman posted 11/19/20 11:51 AM
Sean,
I have isolated the problem further so you can easily reproduce it in your environment. I modified the plugin as follows:

[root@vhlppgppx004 plugins]# cat action_RestartMQIPT_Test.sh
#!/bin/bash
[ $# -ne 0 ] && { echo "Usage: $0"; exit 1; }
/opt/ncpa/plugins/test.sh &
echo "End of plugin" > /tmp/test1
exit $?

And created a test.sh

[root@vhlppgppx004 plugins]# cat test.sh
#!/bin/bash
sleep 1000
exit 0

Now, when I execute the curl command to run the plugin it does not complete but the /tmp/test1 file gets created. Curl completes after test.sh ends after 1000 second sleep.
So running the script in the background prevents NCPA from returning results. It probably waits for all child processes to complete.

Please let me know if you are able to reproduce.



Yakov Korenfeld
Systems Engineer II – Operations Event Management
Infrastructure Delivery
Travelport
e: yakov.korenfeld@travelport.com<mailto:yakov.korenfeld@travelport.com>
m: +1 678-773-5390

**Comment by ccztux** [comment]
I still havent tested it, but i think there is a explanation for it.
1. NCPA has a plugin timeout of 60 seconds. After this 60 seconds the plugin process will be killed (you get no result)
2. As described here:
 >During testing i found another issue with subprocesses opened via ncpa. If a plugin starts some background jobs, these background jobs can not be killed via Popen.kill() method. I will provide a fix for this via a pull request.
i found an issue during testing, that the kill signal will not be propagated to child processes (the background jobs) started in a plugin.

Can you please provide the following informations:
1. ```NCPA``` version
2. ```OS version``` on which NCPA is running

**Comment by ccztux** [comment]
I have already tested it and i confirm this issue. It seems to be like described [here.](https://github.com/NagiosEnterprises/ncpa/issues/714#issuecomment-733834996)

When the plugin runs into the timeout of 60s you should get a message like that:
```
[root@centos7_ncpa_build listener]# /root/git-repos/ncpa/client/check_ncpa.py -H 10.0.0.76 -t mytoken -M 'plugins/action_RestartMQIPT_Test.sh'
UNKNOWN: Execution exceeded timeout threshold of 60s
```

**Comment by corkyman1** [comment]
It is not a timeout issue and I am not getting the message you described because I am not running a Nagios check. I am running a curl command to run the plugin. This is not a ‘normal monitoring’; it is done to trigger a recovery action on specific conditions.

I’ve tried it on 2.1.6 and then after Sean’s suggestion on the latest ncpa agent.
I am running on Linux hrel6

Yakov Korenfeld
Systems Engineer II – Operations Event Management
Infrastructure Delivery
Travelport
e: yakov.korenfeld@travelport.com<mailto:yakov.korenfeld@travelport.com>
m: +1 678-773-5390

**Comment by ccztux** [comment]
What are you expecting regarding the ```/tmp/test1``` file?
>Now, when I execute the curl command to run the plugin it does not complete but the /tmp/test1 file gets created.

What output of the curl command are you expecting? Did you actually get no output from the API call via curl?

**Comment by ccztux** [comment]
It is reproduceable.

There are 2 issues:
1. When the plugin will be killed after reaching the ```plugin_timeout``` the kill signal will not be propagated to childs started from the plugin. The leftover childs are keeping the curl connection open
2. The error message in case of a timeout will not be displayed

I am working on it.

**Comment by corkyman1** [comment]
1.  I am expecting the output of the /tmp/test1 to contain the verbiage ‘End of plugin’. It does happen, so to me it is a prove that plugin actually gets a return from the user script and itself ends. The problem is that the plugin ending does not translate into return to curl because ncpa agent sees other child processes (this is my interpretation).
  2.  I am expecting the output of the curl command to contain the json with rc and any script output that is produced. It actually happens when the background process ends (sleep ends) and the curl returns all that. If the background process keeps running, like in my production case, then curl never returns.  So I actually get no output from the API call via curl.

Yakov Korenfeld
Systems Engineer II – Operations Event Management
Infrastructure Delivery
Travelport
e: yakov.korenfeld@travelport.com<mailto:yakov.korenfeld@travelport.com>
m: +1 678-773-5390

**Comment by corkyman1** [comment]
I think the case that I submitted is still a different flavor of a problem.

Yakov Korenfeld
Systems Engineer II – Operations Event Management
Infrastructure Delivery
Travelport
e: yakov.korenfeld@travelport.com<mailto:yakov.korenfeld@travelport.com>
m: +1 678-773-5390

**Comment by ccztux** [comment]
I have changed your ```test.sh``` script like that:
```
[root@centos7_ncpa_test src]# cat /usr/local/ncpa/plugins/test.sh 
#!/bin/bash

i="0"

while [ "${i}" -le "5" ]
do
	echo "i: $i"
	sleep 1
	i="$(( i + 1 ))"
done
#sleep 1000
exit 0
```

NCPA version 2.2.2:
This is my curl output (**plugin_timeout set to 10 seconds**):
```
[root@centos7_ncpa_build ncpa]# curl -vk "https://10.0.0.85:5693/api/plugins/action_RestartMQIPT_Test.sh/?token=mytoken&check=1"
* About to connect() to 10.0.0.85 port 5693 (#0)
*   Trying 10.0.0.85...
* Connected to 10.0.0.85 (10.0.0.85) port 5693 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* skipping SSL peer certificate verification
* SSL connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate:
* 	subject: CN=centos7,OU=Development,O="Nagios Enterprises, LLC",L=St. Paul,ST=Minnesota,C=US
* 	start date: Aug 19 19:23:50 2020 GMT
* 	expire date: Aug 17 19:23:50 2030 GMT
* 	common name: centos7
* 	issuer: CN=centos7,OU=Development,O="Nagios Enterprises, LLC",L=St. Paul,ST=Minnesota,C=US
> GET /api/plugins/action_RestartMQIPT_Test.sh/?token=mytoken&check=1 HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 10.0.0.85:5693
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 76
< Access-Control-Allow-Origin: *
< X-Frame-Options: SAMEORIGIN
< Content-Security-Policy: frame-ancestors 'self'
< Date: Fri, 27 Nov 2020 19:16:07 GMT
< 
{
    "returncode": 0, 
    "stdout": "i: 0\ni: 1\ni: 2\ni: 3\ni: 4\ni: 5"
* Connection #0 to host 10.0.0.85 left intact
```


This is my curl output (**plugin_timeout set to 5 seconds**):
```
[root@centos7_ncpa_build ncpa]# time curl -vk "https://10.0.0.85:5693/api/plugins/action_RestartMQIPT_Test.sh/?token=mytoken&check=1"
* About to connect() to 10.0.0.85 port 5693 (#0)
*   Trying 10.0.0.85...
* Connected to 10.0.0.85 (10.0.0.85) port 5693 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* skipping SSL peer certificate verification
* SSL connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate:
* 	subject: CN=centos7,OU=Development,O="Nagios Enterprises, LLC",L=St. Paul,ST=Minnesota,C=US
* 	start date: Aug 19 19:23:50 2020 GMT
* 	expire date: Aug 17 19:23:50 2030 GMT
* 	common name: centos7
* 	issuer: CN=centos7,OU=Development,O="Nagios Enterprises, LLC",L=St. Paul,ST=Minnesota,C=US
> GET /api/plugins/action_RestartMQIPT_Test.sh/?token=mytoken&check=1 HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 10.0.0.85:5693
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 76
< Access-Control-Allow-Origin: *
< X-Frame-Options: SAMEORIGIN
< Content-Security-Policy: frame-ancestors 'self'
< Date: Fri, 27 Nov 2020 19:18:36 GMT
< 
{
    "returncode": 0, 
    "stdout": "i: 0\ni: 1\ni: 2\ni: 3\ni: 4\ni: 5"
* Connection #0 to host 10.0.0.85 left intact
}
```
As you can see, the childs will not be killed if plugin_timeout was reached. Curl gets the output of test.sh although the plugin action_RestartMQIPT_Test.sh was killed from ncpa.



NCPA with my patch:
This is my curl output (**plugin_timeout set to 10 seconds**):
```
[root@centos7_ncpa_build ncpa]# curl -vk "https://10.0.0.85:5693/api/plugins/action_RestartMQIPT_Test.sh/?token=mytoken&check=1"
* About to connect() to 10.0.0.85 port 5693 (#0)
*   Trying 10.0.0.85...
* Connected to 10.0.0.85 (10.0.0.85) port 5693 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* skipping SSL peer certificate verification
* SSL connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate:
* 	subject: CN=centos7,OU=Development,O="Nagios Enterprises, LLC",L=St. Paul,ST=Minnesota,C=US
* 	start date: Aug 19 19:23:50 2020 GMT
* 	expire date: Aug 17 19:23:50 2030 GMT
* 	common name: centos7
* 	issuer: CN=centos7,OU=Development,O="Nagios Enterprises, LLC",L=St. Paul,ST=Minnesota,C=US
> GET /api/plugins/action_RestartMQIPT_Test.sh/?token=mytoken&check=1 HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 10.0.0.85:5693
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 76
< Access-Control-Allow-Origin: *
< X-Frame-Options: SAMEORIGIN
< Content-Security-Policy: frame-ancestors 'self'
< Date: Fri, 27 Nov 2020 19:09:47 GMT
< 
{
    "returncode": 0, 
    "stdout": "i: 0\ni: 1\ni: 2\ni: 3\ni: 4\ni: 5"
* Connection #0 to host 10.0.0.85 left intact
```

This is my curl output (**plugin_timeout set to 5 seconds**):
```
[root@centos7_ncpa_build ncpa]# curl -vk "https://10.0.0.85:5693/api/plugins/ction_RestartMQIPT_Test.sh/?token=mytoken&check=1"
* About to connect() to 10.0.0.85 port 5693 (#0)
*   Trying 10.0.0.85...
* Connected to 10.0.0.85 (10.0.0.85) port 5693 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* skipping SSL peer certificate verification
* SSL connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate:
* 	subject: CN=centos7,OU=Development,O="Nagios Enterprises, LLC",L=St. Paul,ST=Minnesota,C=US
* 	start date: Aug 19 19:23:50 2020 GMT
* 	expire date: Aug 17 19:23:50 2030 GMT
* 	common name: centos7
* 	issuer: CN=centos7,OU=Development,O="Nagios Enterprises, LLC",L=St. Paul,ST=Minnesota,C=US
> GET /api/plugins/action_RestartMQIPT_Test.sh/?token=mytoken&check=1 HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 10.0.0.85:5693
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 145
< Access-Control-Allow-Origin: *
< X-Frame-Options: SAMEORIGIN
< Content-Security-Policy: frame-ancestors 'self'
< Date: Fri, 27 Nov 2020 19:10:56 GMT
< 
{
    "returncode": -1, 
    "stdout": "Error: Plugin command (/bin/sh /usr/local/ncpa/plugins/action_RestartMQIPT_Test.sh) timed out. (4 sec)"
* Connection #0 to host 10.0.0.85 left intact
```

With my patch all childs were killed and curl shows the errormessage.

**Comment by jomann09** [comment]
Closing this since I do believe this is fixed in 2.3 with the patch.
